Save message properties into @metadata fields

magnusbaeck opened the pull request:
Apart from the message payload itself, RabbitMQ messages contain various metadata when they're delivered to a consumer. These can include the timestamp the message was published, the name of the user that published the message, the name of the exchange/queue to which the message was published, etc.

Properties of the message and its delivery are stored in the `[@metadata][rabbitmq_properties]` field while the user-defined headers (technically a part of the properties) are stored in the `[@metadata][rabbitmq_headers]` field.

I can't for the life of me find where the preamble part of the documentation ("[Pull events from a RabbitMQ exchange. The default settings will create ...](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-rabbitmq.html)") is stored. It should be updated to include information about these new `@metadata` fields.

This PR addresses both #19 and #24.

elasticsearch-release commented:
Jenkins standing by to test this. If you aren't a maintainer, you can ignore this comment. Someone with commit access, please review this and clear it for Jenkins to run; then  say 'jenkins, test it'.

magnusbaeck commented:
The original commit [failed in Travis](https://travis-ci.org/logstash-plugins/logstash-input-rabbitmq/builds/107502789) with the following problems:

* A ChannelAlreadyClosed exception caused by the insufficient one-second sleep in spawn_and_wait(). I don't quite understand why that sleep is necessary, but I had to prolong it when testing on my machine to get any kind of reliability in the integration tests. Since it's apparently not just my machine I've added commit e4cfa52 that prolongs the sleep. If you're not happy about that please do something about the test.
* The second seemingly unrelated failure is caused by the first one since we're reusing the same queue for all tests (so if a test publishes a message but bails out before it can consume it the next test will get the message from the preceding test). I've addressed that in commit ac36d88.

magnusbaeck commented:
Ping @logstash-plugins/logstash-core—this one's been untouched for a while now.

andrewvc commented on lib/logstash/inputs/rabbitmq.rb:
```diff
@@ -131,17 +139,87 @@ def consume!
         close_connection
       end
 
+      public
```
Can you nix the `public` declarations? That style has fallen by the wayside for LS plugins.

magnusbaeck replied:
What is the preferred style? Put all private methods at the end of the class and precede that group of methods with a single `private`?

andrewvc commented on lib/logstash/inputs/rabbitmq.rb:
```diff
@@ -131,17 +139,87 @@ def consume!
+        #
+        # List values are java.util.ArrayList objects and we need to
+        # recurse into them to convert any nested strings values.
+        if value.respond_to? "toArray"
```
I'm a bit worried about the cost of `#respond_to?` against a native Java object in JRuby. I'm wondering if there's a way to do this either without reflection or with a simple class equivalence check.

Thoughts? /cc @jordansissel @colinsurprenant @guyboertje

ph replied:
I think checking agains't the class will be faster than the `respond?`

jordansissel replied:
I don't know what the performance costs are, but my instinct is that a class check is quicker than reflection. I also feel that a class check is more explicit about the intent of behavior and expectations, but others can feel differently ;)

andrewvc replied:
@magnusbaeck would you mind converting this to a check a check for java.util.ArrayList ? I concur with @jordansissel and @ph here.

andrewvc commented on lib/logstash/inputs/rabbitmq.rb:
```diff
@@ -131,17 +139,87 @@ def consume!
+
+      private
+      def get_headers(metadata)
+        if not metadata.headers.nil?
```
I apologize. We really need to update our style guide. Would you mind using the [higher precedence](http://stackoverflow.com/questions/27429711/can-someone-explain-not-vs-in-ruby) `!` instead of `not`? 

The lower precedence operators obviously work fine, but they can be a subtle source of bugs if someone writes `if not something && something else`.

magnusbaeck replied:
Had no idea. Thanks!

andrewvc commented on lib/logstash/inputs/rabbitmq.rb:
```diff
@@ -131,17 +139,87 @@ def consume!
+      private
+      def get_headers(metadata)
+        if not metadata.headers.nil?
+          Hash[metadata.headers.map {|k, v| [k, get_header_value(v)]}]
```
I like this trick! Lazy evaluation for the win :). That being said, I am concerned with this pattern because it breaks `Hash#keys`, the object might look empty, but in-fact not be. Your thougths @magnusbaeck ?

magnusbaeck replied:
Sorry, my limited Ruby-fu isn't following along. Isn't this a plain old Hash constructor? I thought lazy evaluation only took place with a block constructor.

```
$ irb
jruby-1.7.23 :002 > Hash[{"a" => "b", "c" => "d"}.map {|k, v| [k, v.upcase]}].keys
 => ["a", "c"]
```

andrewvc commented on lib/logstash/inputs/rabbitmq.rb:
```diff
@@ -131,17 +139,87 @@ def consume!
+        #
+        # Technically the exchange, redeliver, and routing-key
+        # properties belong to the envelope and not the message but we
+        # ignore that distinction here. However, we extract the
```
Good call on ignoring that distinction :)

andrewvc commented on lib/logstash/inputs/rabbitmq.rb:
```diff
@@ -131,17 +139,87 @@ def consume!
+          # The AMQP 0.9.1 timestamp field only has second resolution
+          # so storing milliseconds serves no purpose and might give
+          # the incorrect impression of a higher resolution.
+          "timestamp"        => lambda {|x| x.getTime() / 1000 },
```
Since we only have one field with a special case have considered reformatting this as:

```ruby
["app-id", "cluster-id", ...].reduce({}) do |acc,name|
  value = metadata.send(name.gsub("-", "_"))
  if value
    acc[name] = name != "timestamp" ? value : value.getTime / 1000
  end
  acc
end
```

andrewvc commented on lib/logstash/inputs/rabbitmq.rb:
```diff
@@ -131,17 +139,87 @@ def consume!
+        }.each_pair{|name, transform|
+          # The method names obviously can't contain hyphens.
+          value = metadata.send(name.gsub("-", "_"))
+          result[name] = transform.call(value) if not value.nil?
```
I'd like to make the same comment here about low precedence operators I made above.

andrewvc commented on spec/inputs/rabbitmq_spec.rb:
```diff
@@ -166,21 +166,73 @@ def spawn_and_wait(instance)
 
   describe "receiving a message with a queue specified" do
-    let(:queue_name) { "foo_queue" }
+    let(:queue_name) { "logstash-input-rabbitmq-#{rand(0xFFFFFFFF)}" }
```
Not a bad idea!

andrewvc commented on spec/inputs/rabbitmq_spec.rb:
```diff
@@ -166,21 +166,73 @@ def spawn_and_wait(instance)
+      expect(event).to include("@metadata")
+      expect(event["@metadata"]).to include("rabbitmq_headers")
+      expect(event["@metadata"]["rabbitmq_headers"]).to eq({})
+    end
```
Thanks for writing these great tests! :+1:

andrewvc commented on spec/inputs/rabbitmq_spec.rb:
```diff
@@ -166,21 +166,73 @@ def spawn_and_wait(instance)
 
   describe LogStash::Inputs::RabbitMQ do
-    let(:config) { super.merge("queue" => "foo_queue") }
+    # With auto_delete this example just hangs
```
Hmmmm, I'd like to dive into the reason why. Did you notice what was going on here?

It's not a requirement for accepting this patch, but it'd be good to know.

magnusbaeck replied:
No, sorry. I never managed to figure out how to enable logs when running under RSpec so I was blind to what was going on and didn't know how to debug it.

andrewvc replied:
@magnusbaeck you can enable logs by setting the env var `TEST_DEBUG=true` . Mind trying that out?

magnusbaeck replied:
Ah, _that's_ the magic trick. With that I found that the problem simply was that there was a foo_queue with auto_delete disabled, and attempting to create that queue created a conflict. Using an autocreated queue name solved that.

andrewvc commented:
Hi @magnusbaeck , apologies for ignoring this. Things have been crazy in the leadup to Elasticon.

Thanks for the excellent patch! I left some feedback on it for your review.

magnusbaeck commented:
@andrewvc, please have another look. There are still some open follow-up questions from me from last week but I made educated guesses and rebased my patches.

magnusbaeck commented:
Oh, and if you can tell me where the plugin documentation's source is I can update it to include information about the new metadata fields.

andrewvc commented:
Thanks @magnusbaeck , I'm just trying to get some more perf feedback about the one line from @jordansissel , @ph or someone else. Other than that looks great. It looks like the docgen broke for this plugin. The docs should be pulled from comments above the main class, but https://www.elastic.co/guide/en/logstash/current/plugins-inputs-rabbitmq.html looks like its from a very old version. Paging @suyograo to take a look. Any ideas?

andrewvc commented:
@magnusbaeck aside from the class check issue this is good to merge / release. I think we may make docs a separate PR given the breakage in docgen I currently see.

jordansissel commented on lib/logstash/inputs/rabbitmq.rb:
```diff
@@ -135,6 +137,60 @@ def shutdown_consumer
+        # ignore that distinction here. However, we extract the
+        # headers separately via get_headers even though the header
+        # table technically is a message property.
+        ["app-id", "cluster-id", "consumer-tag", "content-encoding",
```
Can you move this array to be a constant?

andrewvc replied:
If we do this you should call '.freeze' on all the strings since a user messing with `@metadata` could modify the global instances.

jordansissel commented on lib/logstash/inputs/rabbitmq.rb:
```diff
@@ -135,6 +137,60 @@ def shutdown_consumer
+            # The AMQP 0.9.1 timestamp field only has second resolution
+            # so storing milliseconds serves no purpose and might give
+            # the incorrect impression of a higher resolution.
+            result[name] = name != "timestamp" ? value : value.getTime / 1000
```
Feels odd to use a ternary for this special case. Maybe a separate call `result["timestamp"] = metadata.timestamp` outside of the loop instead?

andrewvc replied:
Hah, I actually recommended it to @magnusbaeck in a previous comment :)  That said I'm fine with either approach.

magnusbaeck replied:
Putting it outside the loop would mean it should be removed from the (upcoming) constant containing the list of properties and be a special case altogether. That seems like the least preferable option to me. I still think my original solution of a hash with lambdas was the most elegant one :-)

jordansissel commented on lib/logstash/inputs/rabbitmq.rb:
```diff
@@ -135,6 +137,60 @@ def shutdown_consumer
+         "content-type", "correlation-id", "delivery-mode",
+         "exchange", "expiration", "message-id", "priority",
+         "redeliver", "reply-to", "routing-key", "timestamp",
+         "type", "user-id"].reduce({}) do |acc, name|
```
You're using `#reduce` here, but don't do anything with `acc` - is that intentional? If so, you can just use map.

Alternately, you could call `[...].reduce({}) do |result, name| ...` and store into `result[name]` and yield `result` from each block.

magnusbaeck replied:
While extracting the properties into a constant I also noticed that `acc` wasn't used. I'll fix that and get rid of `result`.

jordansissel commented on spec/inputs/rabbitmq_spec.rb:
```diff
@@ -166,21 +166,73 @@ def spawn_and_wait(instance)
     end
+
+    it "should save message properties into a @metadata field" do
+      epoch = 1454789198
```
Is there something special about this time value? or can it be randomized?

magnusbaeck replied:
Nope, it can totally be randomized. Should rand() calls be done in a special way to support Rspec's seed setting?

jordansissel replied:
no special way needed. My understanding is that RSpec seeds random at startup and all you do is just call rand() or similar like you would normally.

jordansissel commented on spec/inputs/rabbitmq_spec.rb:
```diff
@@ -166,21 +166,73 @@ def spawn_and_wait(instance)
+      epoch = 1454789198
+      # Don't test every single property but select a few with
+      # different characteristics to get sufficient coverage.
+      queue.publish("",
```
My preference is that these setup-style activities belong in a `before` block. Example:

```
context "when message properties are available" do
  before do
    epoch = ...
    queue.publish(...)
  end

  let(:event) { output_queue.pop }

  it "should save these properties to @metadata field" do
    expect(event).to include("@metadata")
    ...
 end
end
```

Bonus that this lets you refactor to use shared examples so you can assert that all common behavior exists for all tests (like event has `@metadata` or something)

magnusbaeck commented:
@andrewvc & @jordansissel, I believe all comments have been addressed now so please have another look. I introduced two new commits to a) clean up the existing tests according to Jordan's preference of pushing setup stuff to a `before` block and b) use the plain codec in the tests.

andrewvc commented:
Thanks @magnusbaeck this look stellar! LGTM

magnusbaeck commented:
Thanks @andrewvc! Don't forget to close #19 too. Can you guys prepare a release of this plugin reasonably soon?

suyograo commented:
@magnusbaeck can we get a changelog PR and you can also bump the version in the same PR. we can then publish a new version

magnusbaeck commented:
No problem @suyograo. Assuming a bump to 3.2.0 since it's a backwards-compatible feature addition. What about the documentation? As I said earlier I have no idea where its source is.

andrewvc commented:
@magnusbaeck the doc source should be right above the LogStash::Inputs::RabbitMQ class declaration. The issue is that the doc gen script is currently broken :( . @suyograo do you want to fix that before we publish? It could be.... involved. Could this have been broken since way back when I added the connection mixin?

suyograo commented:
@andrewvc I pushed a new doc update last week. When we release a new version of this plugin, it'll get updated

suyograo commented:
@magnusbaeck yes, 3.2.0 and an entry in https://github.com/logstash-plugins/logstash-input-rabbitmq/blob/master/CHANGELOG.md

andrewvc commented:
@magnusbaeck @suyograo so the doc gen appears to be working. @magnusbaeck would you mind including the docs in your next PR? See https://github.com/logstash-plugins/logstash-input-file/blob/master/lib/logstash/inputs/file.rb for an example of where docs go.

magnusbaeck commented:
> the doc source should be right above the LogStash::Inputs::RabbitMQ class declaration.

Yeah, I know where the docs should go. I was confused by the lack of docs in the rabbitmq.rb found at the tip of the branch combined with the existence of docs in the online edition. But looking at the [2.2 online docs](https://www.elastic.co/guide/en/logstash/2.2/plugins-inputs-rabbitmq.html) there are indeed no introductory paragraphs but they're still around in the [2.1 docs](https://www.elastic.co/guide/en/logstash/2.1/plugins-inputs-rabbitmq.html). Those paragraphs were apparently lost in commit 6bf115a72f907b971d7e5a821abbb206516f67b9, released in v3.0.0 which was included in LS 2.2.

Okay, so I'll send a PR that reinstates the old documentation (with improvements for the new `@metadata` fields) and updates CHANGELOG.md and the gemspec for the 3.2.0 release. I won't have time tonight though.

magnusbaeck commented:
Changelog, version bump, and documentation fixed in #70.

suyograo commented:
Published 3.2.0
